server/routerlicious/packages/lambdas/src/alfred/index.ts

```diff
@@ -667,6 +762,62 @@ export function configureWebSocketServices(
 				})
 				.finally(() => {
 					connectDocumentComplete = true;
+					if (disconnectDocumentP) {
+						Lumberjack.warning(
+							`ConnectDocument completed after disconnect was handled.`,
+						);
+						// We have already received disconnect for this connection.
+						disconnectDocumentP.finally(() => {
+							// We might need to re-run disconnect handler after previous disconnect handler completes.
+							// DisconnectDocument internally handles the cases where we have already run disconnect for
+							// roomsMap and connectionsMap so that we don't duplicate disconnect efforts.
+							// The primary need for this retry is when we receive "disconnect" in the narrow window after
+							// "connect_document" but before "connectDocumentP" is defined.
+							const alreadyDisconnectedAllConnections: boolean =
+								connectionsMap.size === clientIdConnectionsDisconnected.size;
+							const alreadyDisconnectedAllClients: boolean =
+								roomMap.size === clientIdClientsDisconnected.size;
+							if (
+								alreadyDisconnectedAllConnections &&
+								alreadyDisconnectedAllClients
+							) {
+								// Don't retry disconnect if all connections and clients are already handled.
```

Comment:
there could still be a case that connectionsMap and RoomMap is cleared but clientId is still present in redis cache. I think we need to check that too before returning here?

Reply:
I updated the `clientIdClientsDisconnected` setter to only happen if Redis removal is a success.